components/accounts/src/lib.rs

```diff
@@ -0,0 +1,349 @@
+    Init { state_json: Option<String> },
+    CompletedAuthFlow { code: String, state: String },
+    AuthenticationError,
+    Logout,
+    // State-machine generated events.
+    AccountCreated,
+    AccountRestored,
+    RecoveredFromAuthenticationProblem,
+    FailedToAuthenticate,
+}
+
+// State machine states.
+#[derive(Debug, Clone, Copy)]
+pub enum State {
+    Start,
+    NotAuthenticated,
+    AuthenticationProblem,
+    Authenticated,
+}
+
+pub struct FxAccountManager {
+    remote_config: FxAConfig,
+    state: State,
+    account: FirefoxAccount,
+    device_config: DeviceConfig,
+    profile: Option<Profile>,
+    devices: Vec<Device>,
+    // app_scopes: Vec<String>, TODO
+}
+
+impl FxAccountManager {
+    pub fn new(remote_config: FxAConfig, device_config: DeviceConfig) -> Self {
+        let fxa_config = remote_config.clone();
+        Self {
+            remote_config,
+            state: State::Start,
+            account: FirefoxAccount::with_config(fxa_config),
+            device_config,
+            profile: None,
+            devices: vec![],
+        }
+    }
+
+    pub fn init(&mut self, state_json: Option<&str>) {
+        self.process_event(Event::Init {
+            state_json: state_json.map(|s| s.to_owned()),
+        });
+    }
+
+    pub fn begin_oauth_flow(&mut self) -> Result<String> {
+        // Can't really return values from state machine pumping, so here we go instead.
+        self.account.begin_oauth_flow(&[scopes::PROFILE]) // TODO app_scopes
+    }
+
+    pub fn begin_pairing_flow(&mut self, pairing_url: &str) -> Result<String> {
+        self.account
+            .begin_pairing_flow(pairing_url, &[scopes::PROFILE]) // TODO app_scopes
+    }
+
+    pub fn finish_authentication_flow(&mut self, code: &str, state: &str) {
+        self.process_event(Event::CompletedAuthFlow {
+            code: code.to_owned(),
+            state: state.to_owned(),
+        });
+    }
+
+    pub fn get_profile(&mut self) -> Option<Profile> {
+        self.profile.clone()
+    }
+
+    pub fn update_profile(&mut self) -> Option<Profile> {
+        let new_profile = match self.account.get_profile(false) {
+            Ok(p) => Some(p),
+            Err(e) => match e.kind() {
+                ErrorKind::RemoteError { code: 401, .. } => {
+                    self.state = State::AuthenticationProblem;
+                    None
+                }
+                _ => None,
+            },
+        };
+        self.profile = new_profile;
+        self.profile.clone()
+    }
+
+    // Called by Sync or for any third-party service that uses an OAuth token.
+    pub fn on_authentication_error(&mut self) {
+        self.process_event(Event::AuthenticationError)
+    }
+
+    pub fn logout(&mut self) {
+        self.process_event(Event::Logout)
+    }
+
+    pub fn account_state(&self) -> State {
+        self.state
+    }
+
+    pub fn export_persisted_state(&self) -> Result<String> {
+        self.account.to_json()
```

Comment:
This is somehow quite surprising to me...we're going to all this effort to track what state we're in, but we don't actually persist that information to disk. Would it be useful to persist our own state here as well as the underlying fxa-client object?